api/v1beta1/openstackdataplanerole_types.go

```diff
@@ -82,3 +85,41 @@ func init() {
 func (instance OpenStackDataPlaneRole) IsReady() bool {
 	return instance.Status.Conditions.IsTrue(condition.ReadyCondition)
 }
+
+// Validate - validates the shared data between role and nodes
+func (instance OpenStackDataPlaneRole) Validate(nodes []OpenStackDataPlaneNode) error {
+	var errorMsgs []string
+
+	if len(instance.Spec.OpenStackAnsibleEERunnerImage) > 0 || len(instance.Spec.NetworkAttachments) > 0 {
+		for _, node := range nodes {
+			errorMsgs = AssertNodeAndRoleUniqueness(node, instance, errorMsgs)
+		}
+	}
+
+	// Compare nodes when role fields are empty
+	if len(instance.Spec.OpenStackAnsibleEERunnerImage) == 0 || len(instance.Spec.NetworkAttachments) == 0 {
+		nodeMap := make(map[string]OpenStackDataPlaneNode)
+
+		for _, node := range nodes {
+			if len(nodeMap["openStackAnsibleEERunnerImage"].Name) > 0 && len(node.Spec.OpenStackAnsibleEERunnerImage) > 0 && node.Spec.OpenStackAnsibleEERunnerImage != nodeMap["openStackAnsibleEERunnerImage"].Spec.OpenStackAnsibleEERunnerImage {
```

Comment:
There's still some duplicated logic on lines 104 and 108. We could have two functions in common.go 
`AssertNodeAndRoleUniqueness` and `AssertNodeAndNodeUniqueness` where the later is passed two nodes  but compares the same attributes (`RunnerImage` or `NetworkAttachments`).